### PR TITLE
Add the option to use SM for P2P comm in TP overlap

### PR DESCRIPTION
**Pull request by erhoo82**
# Description

Add an option for SM-based P2P send/recv in TP communication overlap.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Changes

Please list the changes introduced in this PR:

- Add an option to select between CE- and SM-based P2P send/recv in TP communication overlap by `use_ce`.
- Update the argument names in userbuffer interface to avoid naming conflicts

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/module/base.py [hunk 1]
```diff
@@ -113,6 +113,7 @@ def add_ub(
         num_splits: int = 0,
         aggregate: int = 0,
         atomic_gemm: int = 0,
+        use_ce: int = 1,
```
Comment (timmoon10) [inline transformer_engine/pytorch/module/base.py:116]: We're a bit inconsistent about the type of `use_ce`. It's `bool` in `tex.UbufP2PCommOverlap`, so it seems appropriate for it to be `bool` here:
```suggestion
        use_ce: bool = True,
```

File: transformer_engine/pytorch/module/base.py [hunk 2]
```diff
@@ -194,12 +196,13 @@ def add_ub(
...
             method = ub_cfg.get("method", get_method(name))
-            num_sm = ub_cfg.get("num_sm", 16)
-            cga_size = ub_cfg.get("cga_size", 2)
+            num_sm = ub_cfg.get("num_sm", 1 if method == "ring_exchange" else 16)
+            cga_size = ub_cfg.get("cga_size", 1 if method == "ring_exchange" else 2)
             num_splits = ub_cfg.get("num_splits", 4 if method == "pipeline" else 0)
             set_sm_margin = ub_cfg.get("set_sm_margin", 0)
             aggregate = ub_cfg.get("aggregate", 0)
             atomic_gemm = ub_cfg.get("atomic_gemm", 0)
+            use_ce = ub_cfg.get("use_ce", 1)
```
Comment (timmoon10) [inline transformer_engine/pytorch/module/base.py:205]:
```suggestion
            use_ce = ub_cfg.get("use_ce", True)
```

**Review by timmoon10 — APPROVED** [review]
Overall looks good, although we're inconsistent between `bool` and `int`.

File: transformer_engine/pytorch/csrc/comm_gemm_overlap.h [hunk 3]
```diff
@@ -211,11 +211,12 @@ struct UbufCommOverlap : torch::CustomClassHolder, UbufBase {
...
-    _ub_comm->sms = comm_sms;
-    _ub_comm->cga_size = cga_size;
-    // Get the current userbuf offset
-    char *ubuf_wt_ptr = reinterpret_cast<char *>(_ubuf.data_ptr());
+    _ub_comm->use_ce = _use_ce;
+    _ub_comm->sms = _num_comm_sm;
+    _ub_comm->cga_size = _cga_size;
+    <
+        // Get the current userbuf offset
+        char *ubuf_wt_ptr = reinterpret_cast<char *>(_ubuf.data_ptr());
```
Comment (timmoon10) [inline transformer_engine/pytorch/csrc/comm_gemm_overlap.h:219]: Merge typo:
```suggestion
    // Get the current userbuf offset
    char *ubuf_wt_ptr = reinterpret_cast<char *>(_ubuf.data_ptr());
```

**Comment by erhoo82** [comment]
@timmoon10 
I think the paddle build failure is unrelated issue?

**Review by denera — APPROVED** [review]
LGTM! I will make sure this also gets into PR #760.

**Pull request merged**